Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -2253,55 +2191,77 @@ void CGContextClearRect(CGContextRef context, CGRect rect) {
                                  nullptr);
     }
 
-    ComPtr<ID2D1Image> currentImage{ image };
+    auto cleanup = wil::ScopeExit([this, layer]() {
+        if (layer) {
+            this->deviceContext->PopLayer();
+        }
 
-    ComPtr<ID2D1Effect> shadowEffect;
-    RETURN_IF_FAILED(_CreateShadowEffect(currentImage.Get(), &shadowEffect));
-    if (shadowEffect) {
-        RETURN_IF_FAILED(shadowEffect.As(&currentImage));
-    }
+        // TODO GH#1194: We will need to re-evaluate Direct2D's D2DERR_RECREATE when we move to HW acceleration.
+        this->PopEndDraw();
+    });
+
+    if (state.HasShadow()) {
```

Comment:
I'd like to see this abstracted out into something like `context.ShouldBuffer()`; additionally:

If you create a CGBitmapContext and pass `data`, we cannot in good faith buffer.
If you create a CGBitmapContext and pass `!data`, we can buffer. **Except**:
    1. If the user calls `CGBitmapContext(Get|Create)Image`, we must flush the buffer before we give them the image.
    2. If the user calls `CGBitmapContextGetData`, we must flush the buffer _and turn off future buffering_.

We need to make sure that anybody who hasn't signed a buffering contract with us gets immediate mode rendering, to most accurately mimic the reference platform.

We might want to consider an expanded version of `CGBitmapContextCreate(...)` or a custom `CGBitmapContextSet[WOC]BufferingMode()` that allows the user to turn _on_ buffering by agreeing to our contract (must `Flush`)